src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/charts/utils.ts

```diff
@@ -54,3 +55,21 @@ export function getDatasourceLayers(
     ...(indexPatternLayers && Object.keys(indexPatternLayers).length ? indexPatternLayers : {}),
   };
 }
+
+export function getLensStateLayer(
+  layers: Record<string, Omit<FormBasedLayer, 'indexPatternId'> | TextBasedLayer>,
+  visLayerId: string | undefined
+) {
+  // Filter to keep non-linked layers (layers without linkToLayers or with linkToLayers set to null)
+  // Also keep ES|QL layers with non-empty columns, as old layers persist after chart type switches and have empty columns
+  const mainLayers = Object.entries(layers).filter(
```

Comment:
Should we use this logic on all visualizations? Which visualizations are affected by these linked layers?